From 5f930d0b5a4dfe7695748b506d085832cd6f980b Mon Sep 17 00:00:00 2001
From: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date: Tue, 26 Aug 2014 17:18:55 -0400
Subject: [input] alps: fix rushmore packet decoding

Message-id: <1409073549-31548-5-git-send-email-benjamin.tissoires@redhat.com>
Patchwork-id: 89250
O-Subject: [PATCH RHEL7 04/18] Input: alps - fix rushmore packet decoding
Bugzilla: 1107819
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1107819
Upstream: since v3.17
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7882179
Tested by me on a Toshiba Portege R30 with a v7 Alps touchpad

commit f105e34a4e3ff073b0d96d5f91eb3c6e8dffaa4c
Author: Yunkang Tang <yunkang.tang@cn.alps.com>
Date:   Fri Jul 25 22:29:24 2014 -0700

    Input: alps - fix rushmore packet decoding

    Signed-off-by: Yunkang Tang <yunkang.tang@cn.alps.com>
    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/input/mouse/alps.c b/drivers/input/mouse/alps.c
index fb15c64..e15daea 100644
--- a/drivers/input/mouse/alps.c
+++ b/drivers/input/mouse/alps.c
@@ -559,6 +559,9 @@ static void alps_decode_rushmore(struct alps_fields *f, unsigned char *p,
 {
  alps_decode_pinnacle(f, p, psmouse);
 
+ /* Rushmore's packet decode has a bit difference with Pinnacle's */
+ f->is_mp = !!(p[5] & 0x40);
+ f->fingers = max((p[5] & 0x3), ((p[5] >> 2) & 0x3)) + 1;
  f->x_map |= (p[5] & 0x10) << 11;
  f->y_map |= (p[5] & 0x20) << 6;
 }
-- 
1.7.1